src/Exthost/LanguageInfo.re

```diff
@@ -78,14 +90,34 @@ let getLanguageFromFilePath = (li: t, fp: string) => {
          Some(Oniguruma.OnigRegExp.Match.getText(matches[1]));
        };
      })
-  |> Stdlib.Option.value(~default);
+  |> Stdlib.Option.value(~default=defaultLanguage);
+}
+
+let getLanguageFromFilePath = (li: t, fp: string) => {
+  let fileName = Path.filename(fp);
+  let extension = Utility.Path.getExtension(fp);
+
+  let updateIfDefault = (f, res) =>
+    if (res == defaultLanguage) {
+      f();
+    } else {
+      res;
+    };
+
+  getLanguageFromOniPath(fp)
+  |> updateIfDefault(() => getLanguageFromExtension(li, extension))
+  |> updateIfDefault(() => getLanguageFromFileName(li, fileName));
 };
 
 let getLanguageFromBuffer = (li: t, buffer: Buffer.t) => {
-  switch (Buffer.getFilePath(buffer)) {
-  | None => defaultLanguage
-  | Some(v) => getLanguageFromFilePath(li, v)
-  };
+  let filePath =
+    switch (Buffer.getFilePath(buffer)) {
+    | None => ""
+    | Some(v) => v
+    };
+
+  getLanguageFromFilePath(li, filePath);
+  // TODO: this should then flow into updateIfDefault(firstLineCheck);
```

Comment:
Perfect! We can check if `Buffer.getNumberOfLines` is greater than 0 - and if it is, we can check `Buffer.getLine(0, buffer) |> BufferLine.raw` to get the raw string, and validate against the regular expression.

We can create a test case too, where we could create some buffer like:
```
let testBuffer = Buffer.ofLines([|"A random first line to test against"|])
|> Buffer.setFilePath(Some("file-path.txt"));
```

And validate that `getLanguageFromBuffer` returns the correct thing. 
Probably the trickiest part of testing this is getting a valid `LanguageInfo.t` to use with it - we could load one of our JSON files like:
```
let languageInfo =
Exthost.Extension.Scanner.load(~category=Exthost.Extension.Scanner.Default, "extensions/shellscript/package.json")
|> Option.to_list
|> Exthost.LanguageInfo.ofExtensions
```
Probably the easiest way, right now!

Reply:
Oh, and I forgot to mention, if you are adding tests to `test/Exthost` to validate this - probably the quickest way is to run:
```
esy '@test' exthost --filter LanguageInfo
```
The exthost tests take a while for the ones that actually start up the extension host - this would filter those out and scope it to just the language tests, which would be much quicker and easier to iterate on.